bee-api/bee-rest-api/src/endpoints/routes/api/v2/messages_submit.rs

```diff
@@ -0,0 +1,208 @@
+            )
+            .await
+        }
+    } else {
+        submit_message_json::<B>(
+            serde_json::from_slice(&bytes.to_vec()).map_err(|e| ApiError::InvalidJson(e.to_string()))?,
+            args.clone(),
+        )
+        .await
+    }
+}
+
+pub(crate) async fn submit_message_json<B: StorageBackend>(
+    value: Value,
+    args: ApiArgsFullNode<B>,
+) -> Result<Response, ApiError> {
+    let protocol_version_json = &value["protocolVersion"];
+    let parents_json = &value["parentMessageIds"];
+    let payload_json = &value["payload"];
+    let nonce_json = &value["nonce"];
+
+    // Tries to build a `Message` from the given JSON.
+    // If some fields are missing, it tries to auto-complete them.
+
+    if !protocol_version_json.is_null() {
+        let parsed_protocol_version = u8::try_from(protocol_version_json.as_u64().ok_or(ApiError::BadRequest(
+            "invalid protocol version: expected an unsigned integer < 256",
+        ))?)
+        .map_err(|_| ApiError::BadRequest("invalid protocol version: expected an unsigned integer < 256"))?;
+
+        if parsed_protocol_version != PROTOCOL_VERSION {
+            return Err(ApiError::BadRequest("invalid protocol version"));
+        }
+    }
+
+    let parents: Vec<MessageId> = if parents_json.is_null() {
+        let mut parents = args
+            .tangle
+            .get_messages_to_approve()
+            .await
+            .ok_or(ApiError::ServiceUnavailable(
+                "can not auto-fill parents: no tips available",
+            ))?;
+        parents.sort_by(|a, b| a.as_ref().cmp(b.as_ref()));
+        parents
+    } else {
+        let parents = parents_json.as_array().ok_or(ApiError::BadRequest(
+            "invalid parents: expected an array of message ids",
+        ))?;
+        let mut message_ids = Vec::with_capacity(parents.len());
+        for message_id in parents {
+            let message_id = message_id
+                .as_str()
+                .ok_or(ApiError::BadRequest("invalid parent: expected a message id"))?
+                .parse::<MessageId>()
+                .map_err(|_| ApiError::BadRequest("invalid parent: expected a message id"))?;
+            message_ids.push(message_id);
+        }
+        message_ids
+    };
+
+    let payload = if payload_json.is_null() {
+        None
+    } else {
+        let payload_dto = serde_json::from_value::<PayloadDto>(payload_json.clone())
+            .map_err(|e| ApiError::InvalidJson(e.to_string()))?;
+        Some(Payload::try_from(&payload_dto).map_err(ApiError::InvalidDto)?)
+    };
+
+    let nonce = if nonce_json.is_null() {
+        None
+    } else {
+        let parsed_nonce = nonce_json
+            .as_str()
+            .ok_or(ApiError::BadRequest("invalid nonce: expected an u64-string"))?
+            .parse::<u64>()
+            .map_err(|_| ApiError::BadRequest("invalid nonce: expected an u64-string"))?;
+        if parsed_nonce == 0 { None } else { Some(parsed_nonce) }
```

Comment:
We can remove line 125 now because milestone nonces can be zero now with the latest stardust changes, right?